### PR TITLE
Fixing Socket async connect for connection-less protocols; Fixing and enabling Socket perf tests.

### LINKED ISSUE #4968: UDP Socket.ConnectAsync result in  "An invalid argument was supplied Error Code:10022"
For version RC1

A very simple call failed and throw exception  "An invalid argument was supplied Error Code:10022"
Repro:

```
var c = new UdpClient();
c.Client.ConnectAsync("114.114.114.114", 53).Wait();
```

It works with sync mode:

```
var c = new UdpClient();
c.Client.Connect("114.114.114.114", 53);
```

Let me know if it's not a bug

### PR DESCRIPTION
**Pull request by CIPop**
Simulating an asynchronous Connect (APM/TPL) for connection-less protocols. 
Moving SocketPerformance to Outerloop and enabling the tests without an execution time expectation.

Fixes #4968, #3635

### CODE REVIEW

**Comment by CIPop** [comment]
@davidsh @stephentoub @ericeil @himadrisarkar PTAL

**Comment by CIPop** [comment]
@dotnet-bot test outerloop win10 debug

File: src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 1]
```diff
@@ -5788,6 +5818,22 @@ private static bool DoDnsCallback(IAsyncResult result, MultipleAddressConnectAsy
             return DoMultipleAddressConnectCallback(PostOneBeginConnect(context), context);
         }
 
+        private sealed class ConnectAsyncResult : ContextAwareResult
+        {
+            private EndPoint _endPoint;
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs]: Nit: readonly?

**Comment by stephentoub** [comment]
LGTM

**Comment by CIPop** [comment]
@dotnet-bot test outerloop windows 7 debug

**Comment by CIPop** [comment]
@dotnet-bot test outerloop windows 10 debug

**Comment by CIPop** [comment]
@dotnet-bot test outerloop ubuntu 14.04 debug

File: src/System.Net.Sockets/tests/FunctionalTests/UdpClientTest.cs [hunk 2]
```diff
@@ -76,6 +74,15 @@ public void Client_Idempotent()
             }
         }
 
+        [Fact]
+        public void Connect_Success()
+        {
+            using (var c = new UdpClient())
+            {
+                c.Client.Connect("114.114.114.114", 53);
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/UdpClientTest.cs]: cc: @ericeil 
```
     System.Net.Sockets.Tests.UdpClientTest.Connect_Success [FAIL]
        System.PlatformNotSupportedException : Sockets on this platform are invalid for use after a failed connection attempt, and as a result do not support attempts to connect to multiple endpoints.
        Stack Trace:
              at System.Net.Sockets.Socket.ThrowIfNotSupportsMultipleConnectAttempts()
              at System.Net.Sockets.Socket.Connect(IPAddress[] addresses, Int32 port)
              at System.Net.Sockets.Socket.Connect(String host, Int32 port)
              at System.Net.Sockets.Tests.UdpClientTest.Connect_Success()
```

File: src/System.Net.Sockets/tests/FunctionalTests/UdpClientTest.cs [hunk 3]
```diff
@@ -66,7 +65,6 @@ public void Client_Idempotent()
             }
         }
 
-        [ActiveIssue(4968)]    
         [Fact]
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/UdpClientTest.cs]: cc: @ericeil 
```
     System.Net.Sockets.Tests.UdpClientTest.ConnectAsync_Success [FAIL]
        System.PlatformNotSupportedException : Sockets on this platform are invalid for use after a failed connection attempt, and as a result do not support attempts to connect to multiple endpoints.
        Stack Trace:
              at System.Net.Sockets.Socket.ThrowIfNotSupportsMultipleConnectAttempts()
              at System.Net.Sockets.Socket.BeginConnect(String host, Int32 port, AsyncCallback requestCallback, Object state)
              at System.Net.Sockets.SocketTaskExtensions.<>c.<ConnectAsync>b__5_0(String targetHost, Int32 targetPort, AsyncCallback callback, Object state)
              at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2](Func`5 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, Object state, TaskCreationOptions creationOptions)
              at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2](Func`5 beginMethod, Action`1 endMethod, TArg1 arg1, TArg2 arg2, Object state)
              at System.Net.Sockets.SocketTaskExtensions.ConnectAsync(Socket socket, String host, Int32 port)
              at System.Net.Sockets.Tests.UdpClientTest.<ConnectAsync_Success>d__6.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

**Comment by CIPop** [comment]
@ericeil @stephentoub 
> System.PlatformNotSupportedException : Sockets on this platform are invalid for use after a failed connection attempt, and as a result do not support attempts to connect to multiple endpoints.

This is unexpected for an IP written as a string. As far as I understand, Connect(Async) methods will always try to go through the multiple address code. We should have code handling the special case when a single address is actually available and bypass the PNSE.

I'll track this with a different issue.

**Comment by stephentoub** [comment]
> We should have code handling the special case when a single address is actually available and bypass the PNSE

Which of these are you suggesting:
1. First try to use IPAddress.Parse on the supplied string, and if it's successful, use that address, otherwise fail.
2. Use DNS to resolve the associated endpoints, and if there's only one, use that, otherwise fail.

Eric explicitly decided against (2) for 1.0.  (1) seems reasonable to me, as it should be deterministic (unless of course it's possible for DNS to resolve an IP-address-as-a-host-name to something other than that IP address?)

**Comment by davidsh** [comment]
LGTM

**Comment by CIPop** [comment]
> Which of these are you suggesting:

I think both. The problem with connection-less protocols such as UDP is that the `Connect` is only instructing the TCP stack to bind to the remote endpoint and I don't think it's possible to bind to more than one IP.

Maybe do 1 for everything then 2 only for connection-less protocols?

**Comment by CIPop** [comment]
@dotnet-bot test outerloop Ubuntu 14.04 debug

**Comment by CIPop** [comment]
@dotnet-bot test outerloop Ubuntu debug

**Comment by CIPop** [comment]
@dotnet-bot test outerloop Ubuntu14.04 debug

**Comment by CIPop** [comment]
@dotnet-bot test outerloop OSX debug

**Comment by CIPop** [comment]
@dotnet-bot test Innerloop Windows_NT Debug Build and Test please.

**Comment by CIPop** [comment]
Test Innerloop Windows_NT Debug Build and Test

**Comment by CIPop** [comment]
Outerloop for Unix is failing because of the new Perf tests. I have disabled them against #9189.

**Comment by CIPop** [comment]
Closed & reopened: CI testing was blocked.

**Comment by CIPop** [comment]
Ubuntu14.04 timed out (tracked by #3697):

System.Net.Sockets.Tests.DualMode.AcceptAsyncV6BoundToSpecificV4_CantConnect (from (empty))

Failing for the past 1 build (Since Failed#1876 )
Took 5 sec.
Stacktrace

MESSAGE:
Assert.Throws() Failure\nExpected: typeof(System.Net.Sockets.SocketException)\nActual: typeof(System.TimeoutException): Timed out while waiting for the server accept...
+++++++++++++++++++
STACK TRACE:
at System.Net.Sockets.Tests.DualMode.DualModeConnect_AcceptAsync_Helper(IPAddress listenOn, IPAddress connectTo)

**Comment by CIPop** [comment]
Test Innerloop Ubuntu14.04 Release Build and Test

**Comment by CIPop** [comment]
@mmitche I believe that there's something wrong with the CI system again: the OSX shows as In progress but the [report ](http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/osx_debug_prtest/1946/) shows:

```
Success Build #1946 (Jun 7, 2016 1:29:15 PM)

PR #9137: Fixing Socket async connect...
	Build Artifacts
	build.tar.gz	77.02 MB	 view
	msbuild.log	8.15 MB	 view
	No changes.
	
GitHub pull request #9137 of commit 2571a344320da52acb54f0f6dfe336b4fcb85aed, no merge conflicts.

	Test Result (no failures)
```

Please let me know if it's OK to merge this based on the report above.

**Comment by mmitche** [comment]
@CIPop yeah, we hit the rate limit at the end of the run, and the system was unable to update the Github status.  I'm currently working on this and I think I have a large number of the API hits removed, but I still need additional testing.

**Comment by mmitche** [comment]
(but yes you are good to merge)

**Comment by CIPop** [comment]
Thanks @mmitche!

**Pull request merged**